[webview_flutter] Add webview_flutter_tizen package

seungsoo47 opened the pull request:
* This PR is created from the PR below:
2bfc038 [webview_flutter] Update README and pubspec (#53)
bdd6a15 [webview_flutter] Enable intergation test (#52)
18da2a6 [webview_flutter] Update code by flutter version-up (#49)
1b6eeed [webview_flutter] Handle CompositionEvent (#47)
791511e [webview_flutter] Optimize tbm allocation (#42)
7059ce2 [webview_flutter] Add cookie and scroll APIs (#40)
c5a8930 Add some APIs (#37)
3feae1d [webview_flutter] Add NavigationDelegate and onWebResourceError (#36)
42b63f8 [webview_flutter] Add removeJavascriptChannels API and Callbacks (#31)
749930e [webview_flutter] Update LWE (#28)
9e26386 [webview_flutter] Implement addJavascriptChannels() (#27)
6a34ae0 [webview_flutter] Update LWE (#26)
a438093 [webview_flutter] Add evaluateJavascript() function (#25)
0d2c271 [webview_flutter] Update webview to be focused (#19)
b02b245 Enable IMF on Webview (#17)
571fa4a [webview_flutter_dev] Change webview interface && embedding lwe.so file (#15)
3413964 [webview_flutter_dev] Enable gl compositor at LWE (#14)
e988cf0 Change method name's first letter (#13)
cb8ee9b Handle key events for tizen webview (#11)
44987de [webview_flutter] Add webview_flutter initial package (#7)

Co-authored-by: MuHong Byun <mh.byun@samsung.com>
Co-authored-by: Seungsoo Lee <seungsoo47.lee@samsung.com>

swift-kim commented:
What are those `.so` files? Which source code were they built from?

seungsoo47 commented:
The file is a library for Tizen lightweight web engine. Because the file is not included in Tizen Studio yet, we decided to include them in the plugin.

seungsoo47 commented:
I am sorry, I found unnecessary files. I will remove them.

swift-kim commented:
1. Where is the source code for the library? How can you build it?
2. You provided .so files for the arm architecture only. Isn't this plugin supported on emulators? If not, why?

> the file is not included in Tizen Studio yet

3. Is the library expected to be part of the Tizen SDK? If so, is there any detailed plan?
4. Why is the library necessary for this plugin to function? What makes it differ from the built-in web engine shipped with existing Tizen devices?

seungsoo47 commented:
1. Please refer to link below:
https://review.tizen.org/gerrit/gitweb?p=platform/upstream/lightweight-web-engine.git;a=tree;h=refs/heads/tizen;hb=refs/heads/tizen

2. The library only supports the ARM architecture. Initially, we had implemented General Buffer-based webview, but after that, we changed to FBO-based webview to improve performance.
This guarantees better performance, but there was a problem not running on the emulator. The current solution is not satisfactory for both device and emulator, but we will try to find the solution ASAP.


3. I am not sure. But if the library is not included, one way is to include x86 library in the plugin.

4. We needed to implement a webview that can share memory buffer to draw web pages on the display, also we didn't know how flutter and dart work with built-in webview on devices.
So first we tried to merge the lightweight web engine that we made on Tizen. Later, We will extend to new webview with Tizen builtin web engine based on chromium.

seungsoo47 commented on packages/webview_flutter/tizen/project_def.prop:
```diff
@@ -0,0 +1,30 @@
+# Compiler/linker flags
+USER_CFLAGS_MISC =
+USER_CPPFLAGS_MISC = -c -fmessage-length=0 
+USER_LFLAGS = -llightweight-web-engine.mobile -Wl,-rpath=/opt/usr/home/owner/apps_rw/org.tizen.webview_flutter_tizen_example/lib/arm
```
Is /opt/usr/home/owner/apps_rw/org.tizen.webview_flutter_tizen_example/lib/arm same as $ORIGIN?

swift-kim commented:
Including a ELF file in source tree is not a good practice. Would it be possible to let the user download the file directly from download.tizen.org or the _Releases_ tab of this repo?

Also why does the so file has the `.mobile` suffix in its name? Is the file expected work with any profile (wearable/tv)?

seungsoo47 commented:
I have updated the file's suffix from'.mobile' to'.flutter'.

rwalczyna commented on packages/webview_flutter/tizen/src/webview.h:
```diff
@@ -0,0 +1,69 @@
+  // Key input event
+  virtual void DispatchKeyDownEvent(Ecore_Event_Key* key) override;
+  virtual void DispatchKeyUpEvent(Ecore_Event_Key* key) override;
+  virtual void DispatchCompositionUpdateEvent(const char* str,
+                                              int size) override;
+  virtual void DispatchCompositionEndEvent(const char* str, int size) override;
+
```
When I tried to run example on Tizen device (`flutter-tizen run -v`), two errors occurred:

```
           "/home/r.walczyna/tizen-studio/tools/smart-build-interface/../llvm-10/bin/clang++" -c "src/webview_flutter_tizen_plugin.cc" -o "Debug/src/webview_flutter_tizen_plugin.o" -DTIZEN_DEPRECATION -DDEPRECATION_WARNING
-DTIZEN_DEPRECATION
[...]
           In file included from src/webview_flutter_tizen_plugin.cc:16:
           src/webview.h:34:16: error: 'DispatchCompositionUpdateEvent' marked 'override' but does not override any member functions
             virtual void DispatchCompositionUpdateEvent(const char* str,
                          ^
           src/webview.h:36:16: error: 'DispatchCompositionEndEvent' marked 'override' but does not override any member functions
             virtual void DispatchCompositionEndEvent(const char* str, int size) override;
                          ^
           2 errors generated.
```

After removing not needed `overrides` it has built successful.

seungsoo47 replied:
It seems that you need to update flutter-tizen. Actually, I have updated "flutter_platform_view.h" in flutter engine and the file should been applied on flutter-tizen cache side.

swift-kim replied:
An engine release containing [this patch](https://github.com/flutter-tizen/engine/pull/52) has just been released yesterday. The embedder API versioning is a quite complicated problem because a plugin can have only a constraint on the Flutter SDK version (below) but not the Tizen embedder version. I'll tell you a bit more about this matter later.

```yaml
environment:
  flutter: ">=2.0.1"
```

swift-kim commented on packages/webview_flutter/pubspec.yaml:
```diff
@@ -0,0 +1,24 @@
+name: webview_flutter_tizen
+description: Tizen implementation of the webview plugin
+homepage: https://github.com/flutter-tizen/plugins
+version: 1.0.0
```
Something like `0.1.0` would be better as this is a beta release.

seungsoo47 replied:
Done.

swift-kim reviewed (commented):
I believe libraries that a plugin depends on (`liblightweight-web-engine.flutter.so` in this case) should not be manually copied from the plugin directory into a project directory by the developer. Instead, they should automatically be copied by the flutter-tizen tool into a build output directory directly (this can be done in `TizenPlugins` of [tizen_build_target.dart](https://github.com/flutter-tizen/flutter-tizen/blob/master/lib/tizen_build_target.dart)). I or @HakkyuKim will prepare a fix for this if you agree.

seungsoo47 commented:
> I believe libraries that a plugin depends on (`liblightweight-web-engine.flutter.so` in this case) should not be manually copied from the plugin directory into a project directory by the developer. Instead, they should automatically be copied by the flutter-tizen tool into a build output directory directly (this can be done in `TizenPlugins` of [tizen_build_target.dart](https://github.com/flutter-tizen/flutter-tizen/blob/master/lib/tizen_build_target.dart)). I or @HakkyuKim will prepare a fix for this if you agree.

Thank you for your help. Please correct the copy issue. And I will update results for other issues.

seungsoo47 commented on packages/webview_flutter/.metadata:
```diff
@@ -0,0 +1,10 @@
+# This file tracks properties of this Flutter project.
```
Done.

seungsoo47 commented on packages/webview_flutter/example/.metadata:
```diff
@@ -0,0 +1,10 @@
+# This file tracks properties of this Flutter project.
```
Done.

seungsoo47 commented on packages/webview_flutter/CHANGELOG.md:
```diff
@@ -0,0 +1,3 @@
+## 0.0.1
+
+* TODO: Describe initial release.
```
Done.

seungsoo47 commented on packages/webview_flutter/example/pubspec.yaml:
```diff
@@ -0,0 +1,23 @@
+    path: ../../integration_test/
+
+flutter:
+  uses-material-design: true
```
Done.

seungsoo47 commented on packages/webview_flutter/example/tizen/shared/res/input.html:
```diff
@@ -0,0 +1,16 @@
+			console.log("Hello: " + document.getElementById("ip1").value);
+		}
+	</script>
+</html>
```
This was just for test. So, I removed this file.

seungsoo47 commented on packages/webview_flutter/example/tizen/tizen-manifest.xml:
```diff
@@ -0,0 +1,20 @@
+		<privilege>http://tizen.org/privilege/internet</privilege>
+		<privilege>http://tizen.org/privilege/wifidirect</privilege>
+		<privilege>http://tizen.org/privilege/filesystem.read</privilege>
+		<privilege>http://tizen.org/privilege/filesystem.write</privilege>
```
Done.

swift-kim commented on packages/webview_flutter/example/tizen/tizen-manifest.xml:
```diff
@@ -0,0 +1,13 @@
+    </ui-application>
+	<privileges>
+		<privilege>http://tizen.org/privilege/internet</privilege>
+	</privileges>
```
nit: Indent using spaces.

seungsoo47 replied:
Done.

swift-kim commented on packages/webview_flutter/lib/webview_flutter_tizen.dart:
```diff
@@ -0,0 +1,849 @@
+  bool _initialized = false;
+  FocusNode _focusNode;
+  int _textureId;
+  int get textureId => _textureId;
```
This property is actually never used?

seungsoo47 replied:
I have removed the property.

swift-kim commented on packages/webview_flutter/README.md:
````diff
@@ -0,0 +1,63 @@
+
+To use this plugin, add below lines under the `<manifest>` section in your `tizen-manifest.xml` file,
+
+```xml
+<privileges>
+  <privilege>http://tizen.org/privilege/internet</privilege>
+</privileges>
+
+## Usage
+
+This package is not an _endorsed_ implementation of `webview_flutter`. Therefore, you have to include `webview_flutter_tizen` alongside `webview_flutter` as dependencies in your `pubspec.yaml` file.
+
+```yaml
+dependencies:
+  webview_flutter: ^1.0.6
+  webview_flutter_tizen: ^1.0.0
````
- Broken markdown
- The current version is 0.1.0

seungsoo47 replied:
Done.

swift-kim reviewed (commented):
Please avoid using force-push multiple times as it makes difficult to track code changes. 😢


Theses are just markers for revisiting the original source code later:

- `TizenWebView` was copied and modified from `AndroidWebView` in `package:webview_flutter/src/webview_android.dart`.
  - `TizenView` was copied and modified from `AndroidView` in `package:flutter/src/widgets/platform_view.dart`.
    - `_TizenWebViewState` was copied and modified from `_AndroidViewState` in the same file.
      - `_TizenPlatformTextureView` was copied from `_AndroidPlatformView` in the same file.
        - `RenderTizenView` was copied and modified from `RenderAndroidView` in `package:flutter/src/rendering/platform_view.dart`
          - The enums `PlatformViewHitTestBehavior` and `_PlatformViewState` are from the same file.
          - `_PlatformViewGestureMixin` and `_PlatformViewGestureRecognizer` were copied from the same file.
            - `_factoryTypesSetEquals` and `_factoriesTypeSet` are from the same file.
      - `TizenViewController` was copied and modified from `AndroidViewController` and `TextureAndroidViewController` in `package:flutter/src/services/platform_views.dart`.
        - The enum `_TizenViewState` is the equivalent of `_AndroidViewState` in the same file.
      - `PlatformViewsServiceTizen` was copied and modified from `PlatformViewsService` in `package:flutter/src/services/platform_views.dart`.

Please correct me if anything is wrong.

seungsoo47 commented on packages/webview_flutter/README.md:
```diff
@@ -0,0 +1,52 @@
+  @override
+  void initState() {
+    super.initState();
+    WebView.platform = TizenWebView();
```
I'm sorry but I don't understand what you said. Please let me know what I should do.

swift-kim replied:
Flutter is a cross-platform framework, and the same app code is shared across different platforms. So if you write code like this:
```dart
WebView.platform = TizenWebView();
```
this app will break on platforms other than Tizen, so you need to check which platform the app is currently running on:
```dart
if (Platform.environment.containsKey('TIZEN_API_VERSION')) { // equivalent of `Platform.isAndroid` on Android
  WebView.platform = TizenWebView();
}
```
However this is quite cumbersome for app developers to do, so we can let the plugin itself do its initialization during app startup. Specify `dartPluginClass` as follows in puspec.yaml (it could be anything other than `TizenWebView`):

```yaml
tizen:
  pluginClass: WebviewFlutterTizenPlugin
  fileName: webview_flutter_tizen_plugin.h
  dartPluginClass: TizenWebView
```

and create a static register method in `TizenWebView`:

```dart
class TizenWebView implements WebViewPlatform {
  /// Registers this class as the default instance of [WebViewPlatform].
  static void register() {
    WebView.platform = TizenWebView();
  }
  ...
```

That's it. You can refer to other Dart plugins such as url_launcher_tizen and shared_preferences_tizen which do the similar thing.

swift-kim replied:
I merged https://github.com/flutter-tizen/flutter-tizen/pull/65 so you can test the above code now.

seungsoo47 replied:
I updated it. Thanks.

seungsoo47 commented on packages/webview_flutter/tizen/src/webview_factory.cc:
```diff
@@ -0,0 +1,49 @@
+  std::string localstoragePath =
+      "/tmp/" + std::string("StarFish_localStorage.db");
+  std::string cookiePath = "/tmp/" + std::string("StarFish_cookies.db");
+  std::string cachePath = "/tmp/" + std::string("Starfish_cache.db");
```
I have fixed it.

seungsoo47 commented on packages/webview_flutter/tizen/src/webview.cc:
```diff
@@ -0,0 +1,880 @@
+  float scale_factor = 1;
+
+  webview_instance_ = (LWE::WebContainer*)createWebViewInstance(
+      0, 0, width_, height_, scale_factor, "SamsungOneUI", "ko-KR",
+      "Asia/Seoul",
```
These are default values. Actually, we use results from ICU.

swift-kim replied:
You mean the inputs are ignored by the LWE?

seungsoo47 replied:
If LWE can't get that inputs from ICU, they are used.

seungsoo47 commented on packages/webview_flutter/tizen/src/webview.cc:
```diff
@@ -0,0 +1,880 @@
+        }
+      } else {
+        throw std::invalid_argument("Unknown WebView setting: " + k);
+      }
```
This plugin is in alpha version. As far as I know, the APIs have been added to the latest version. When our plugin is updated to chromium-based webview, APIs may be also updated.

seungsoo47 commented on packages/webview_flutter/tizen/src/webview.cc:
```diff
@@ -0,0 +1,880 @@
+}
+
+static LWE::KeyValue EcoreEventKeyToKeyValue(const char* ecore_key_string,
+                                             bool is_shift_pressed) {
```
I have fixed it. Thanks.

seungsoo47 commented on packages/webview_flutter/tizen/src/webview.cc:
```diff
@@ -0,0 +1,880 @@
+    return LWE::KeyValue::TVColor;
+  } else if (strcmp("XF86PlayBack", ecore_key_string) == 0) {
+    return LWE::KeyValue::TVPlayBack;
+  }
```
Yes. these keys are used in LWE. I think they may be removed in chromium version.

seungsoo47 commented on packages/webview_flutter/tizen/src/webview.cc:
```diff
@@ -0,0 +1,880 @@
+#ifdef TV_PROFILE
+  if ((strncmp(key_name.data(), "XF86Back", 8) == 0)) {
+    key_name = "Escape";
+  }
```
I have fixed it.

seungsoo47 commented on packages/webview_flutter/README.md:
````diff
@@ -0,0 +1,60 @@
+  webview_flutter_tizen: ^0.1.0
+```
+
+To enable tizen implementation, you can refer to the following:
````
I've fixed it.

HakkyuKim commented:
> I believe libraries that a plugin depends on (`liblightweight-web-engine.flutter.so` in this case) should not be manually copied from the plugin directory into a project directory by the developer. Instead, they should automatically be copied by the flutter-tizen tool into a build output directory directly (this can be done in `TizenPlugins` of [tizen_build_target.dart](https://github.com/flutter-tizen/flutter-tizen/blob/master/lib/tizen_build_target.dart)). I or @HakkyuKim will prepare a fix for this if you agree.

See [flutter-tizen/support-binary-dependencies-in-plugin](https://github.com/flutter-tizen/flutter-tizen/pull/68)

HakkyuKim commented:
> > I believe libraries that a plugin depends on (`liblightweight-web-engine.flutter.so` in this case) should not be manually copied from the plugin directory into a project directory by the developer. Instead, they should automatically be copied by the flutter-tizen tool into a build output directory directly (this can be done in `TizenPlugins` of [tizen_build_target.dart](https://github.com/flutter-tizen/flutter-tizen/blob/master/lib/tizen_build_target.dart)). I or @HakkyuKim will prepare a fix for this if you agree.
> 
> Work in progress: [flutter-tizen/support-binary-dependencies-in-plugin](https://github.com/flutter-tizen/flutter-tizen/pull/68)

The feature is merged to master, the tool will now put dependent binaries in tpks. You need to set the `project_def.prop` file as follows:

```
...
USER_LIB_DIRS = lib/${BUILD_ARCH}
USER_LIBS = lightweight-web-engine.flutter
...
```

and place `liblightweight-web-engine.flutter.so` in `tizen/lib/armel`.

Future plugin projects created by the tool will have the following comments written in the `project_def.prop` file:

```
# To provide libraries for your plugin, you must put them in specific
# directories that match their supporting cpu architecture type.
# Due to compatibility issue with Tizen CLI, directory names must be armel for arm 
# and i586 for x86. 
```

seungsoo47 commented:
Thank you. I updated it in accordance with your guidance and I checked that the example worked fine.
